### DISCUSSION TITLE
Migration v4 - context & DataSource

### DISCUSSION THREAD

**Discussion by RemyMachado** [apollographql/apollo-server#7096] [Q&A] [answered]
With v3, I could access any dataSource inside any other dataSource through the context.
> v3 DOC: Apollo Server invokes the dataSource.initialize({ cache, context }) function on each new DataSource

It allowed me to do the following:
```typescript
    // for any DataSource (e.g. Posts)

    async retrieveBearerUser() {
        return this.context.dataSources.users // using users DataSource here
            .findOneById(this.context.userJwtPayload?.subject)
    }
```

But now that the `dataSources` are computed at the same time as the other fields of the context, I don't know how to replicate this behavior.

I had a context looking like a loop thanks to the previous initialization process:
```typescript
dataSources = {
     posts: {
         context: {
             dataSources: {
                 posts: ... // and so on
             }
         }
    }
}
```

It may have been an unintended behavior, so how can I still retrieve the user from another dataSource with the new context implementation?

Any help would be greatly appreciated.

**Comment by glasser** [comment, marked as answer]
There's an example of how to do this [in the migration guide](https://www.apollographql.com/docs/apollo-server/migration/#datasources), specifically under "If you want to access your entire context's value within your DataSource". (It takes advantage of the fact that you can access `this` inside a class's constructor even before the class is fully constructed.)

On the other hand, if there's only one thing like this that you want to pull off the context (eg `userJwtPayload`), another alternative would be to *not* pass the entire `context` into your class and instead just pass the `userJwtPayload` itself in, which removes the circular dependency (and also means that when writing tests for your data source class you don't have to mock out the entire context object!).

**Reply by RemyMachado** [reply]
Thanks for the prompt response and apologies for missing the dedicated part of the guide.
> It takes advantage of the fact that you can access this inside a class's constructor even before the class is fully constructed.

Great, I feel closer to the solution!

___

> your entire context's value within your DataSource

Indeed, it allows to acces the entirer context's value, but now I have another problem.
How can I replicate the async context initialization, based on a promise?

```typescript
// class ContextValue (v4)
    constructor(...) {
        const bearerToken = req.headers.authorization?.split(' ')?.[1]

        if (bearerToken) {
            // promise here
            verifyUserJwt(bearerToken)
                // Add the userJwtPayload to the context
                .then((userJwtPayload) => {
                    this.userJwtPayload = userJwtPayload
                })
        }
        
        this.dataSources = getDataSources({ db, context: this, cache }) // this.userJwtPayload might not have been processed yet
    }
```

since the class constructor can't be async, I'm affraid the `dataSources` will be constructed without the mandatory `context.userJwtPayload`.

___

>  another alternative would be to not pass the entire context into your class and instead just pass the userJwtPayload itself in, which removes the circular dependency

I would be glad to get rid of the circular dependency, but I don't see how passing only the `userJwtPayload` would allow me to retrieve the user (containing private `roles` for authorization) from the DB for any `dataSource`.
The problem lies in the fact that I retrieve the `bearerUser` from the dedicated `dataSources.users`.
Unless I:
- initialize the `user` `dataSource` first
- retrieve the `bearerUser`
- initialize the remaining `dataSource`s that depends on the `this.context.bearerUser`

Furthermore, I'm affraid it would delay the process of every request.

___

Do you see solutions for this two problems?


EDIT:
I forgot that the server can still process an async function for the server `context` field and therefore retrieve the `userJwtPayload` before constructing the `ContextValue` class.
```typescript
expressMiddleware(server, {
            context: async ({ req }) => {
                /* using `req` argument to access headers */

                /* Get the user token from the headers. */
                const bearerToken = req.headers.authorization?.split(' ')?.[1]

                /* skip auth process if no JWT */
                if (!bearerToken) {
                    return new ContextValue({
                        req,
                        server,
                        db,
                        userJwtPayload: undefined,
                    })
                }

                /* authentication process */
                return (
                    verifyUserJwt(bearerToken)
                        // Add the userJwtPayload
                        .then((userJwtPayload) => {
                            return new ContextValue({
                                req,
                                server,
                                db,
                                userJwtPayload,
                            })
                        })
                        // or undefined
                        .catch((error) => {
                            console.info('Error during JWT parsing', error)
                            return new ContextValue({
                                req,
                                server,
                                db,
                                userJwtPayload: undefined,
                            })
                        })
                )
            },
        })
```

**Reply by glasser** [reply]
Yep, my suggestion would be to do the async work first in the `context` function. Does that work well? You should also be able to use `await verifyUserJwt(bearerToken)` if you find that easier to read than then/catch.

**Reply by RemyMachado** [reply]
Yes, it's working well, thanks for your help!

About the `dataSources` inside the context (circular dependency), I found them necessary to make database jointures between collections. I'm not sure I will ever remove them from the context.

**Reply by simplecommerce** [reply]
@glasser This is the issue that I am experiencing also, based on your answer, I know I can add my authentication mechanism here, but my problem is that with this logic, this would mean that any middleware that has to manipulate the context and pass it to the resolvers, would have those manipulations unavailable in the datasources, which can be an issue. 

In my specific case, my authentication mechanism requires datasources access to validate the token and user and then modifies the context to pass it down to the other middlewares and resolvers.

Prior to upgrade to v4, this was working with no issues.
Is there a way to reproduce this like it was in the prior version?

What was the reasoning on why this behavior was changed?

UPDATE: 

Ok so I sort of understand what my issue is now.

I am using `graphql-middleware` to apply middleware resolvers to my schema.
For some reason, the `context` object being passed isn't a reference to the same class instance that I pass to my datasources. 

Which is odd, I opened up a discussion on their hub to see if they can help.

But I was able to fix my issue by creating a middleware that is executed before all my other middlewares.

And I did something like this:

```
export default async (resolve, parent, args, context, info) => resolve(parent, args, context.dataSources.mongo.config.context, info);
```

In this snippet, my middleware grabs the context one of my datasources receives (which is a class instance) and passed that down.

In this way, in my other middlewares if I actually modify the context, it will reflect the changes also in my datasources context instance.

I don't like this but it works until there is a better solution.

**Reply by glasser** [reply]
We found that in practice, Apollo Server's feature of letting you construct the "dataSources" subfield of your context object in a separate function from the construction of the rest of the context led to a lot of confusion:

- It was not obvious that the dataSources function was called once per request rather than once at startup
- It meant that the data type returned from the `context` function was not actually the type of your context object

We felt that making data sources less magical and just normal objects that take their dependencies as constructor arguments was an easier to understand approach. This does mean that if you want a circular reference between your entire context object and one particular field on it (a data source), you have to do a tad more work instead of letting Apollo Server sneakily set it up behind your back — but circular references are subtle enough that it's maybe not a bad thing that you have to do it yourself!

**Reply by simplecommerce** [reply]
Am I correct in assuming that the reason the context object in my middleware is not a reference to the same class instance in my datasources are a result of me using graphql-middleware library and not something underlying with @apollo/server's code?

I am still waiting on feedback on their end.

**Reply by trevor-scheer** [reply]
@simplecommerce we actually do a shallow clone of the context object returned by your context function. If you depend on referential identity you can nest the shared bits of context into an object within your context object. This is understandably an unobvious gotcha.

For some more context:
https://github.com/apollographql/apollo-server/blob/d7e9b97595b063f1e796ec4449850a16d19e8b18/packages/server/src/ApolloServer.ts#L1261-L1276

**Reply by simplecommerce** [reply]
> @simplecommerce we actually do a shallow clone of the context object returned by your context function. If you depend on referential identity you can nest the shared bits of context into an object within your context object. This is understandably an unobvious gotcha.
> 

Ahh ok I kind of understand, if you can could you write me an example it doesn't have to be complex just so I know how to proceed if I want to adapt this on my project?

Thanks for the help I appreciate it!

UPDATE: ok I get what you mean now, I just tested it and confirm it does work as you mentioned. Thanks again!!

**Reply by trevor-scheer** [reply]
Glad that helped, and thanks for the update!
